### DISCUSSION TITLE
Bagaimana Cara merubah Inline string js/json/style jadi punya url/ internal url?

### DISCUSSION THREAD

**Discussion by yuliakbard** [mecha-cms/mecha#266] [Q&A] [answered]
misalnya ane pake inline json f3h yang di hasilkan dari kode dibawah, dijadikan internal url seperti **localhost/manifest.json**. dan juga sama halnya seperti inline js/ css diubah jadi internal url js/ css. apa memungkinkan kang?

```
    'state' => [
        'cache' => false,
        'history' => true,
        // Add custom request header(s) here
        'lot' => [
            'x-requested-with' => 'F3H'
        ],
        'sources' => '[class="noload"], form',
        'turbo' => false,
        'type' => 'document', // Default response type
        'types' => [
            "" => 'document', // Default response type for extension-less URL
            'CS' => 'text',
            'JS' => 'text',
            'JSON' => 'json'
        ]
    ]
```

**Comment by yuliakbard** [comment]
Bagaimana cara merubah rute js/css yang tersimpan di folder layout dan cara menghilangkan filemtime yang ada di urlnya? 

Misal:
`localhost/lot/y/template/index.js` jadi `localhost/index.js`
`localhost/lot/y/template/ads.txt` jadi `localhost/ads.txt`
`localhost/lot/y/template/manifest.json` jadi `localhost/manifest.json`

untuk saat ini ane cuma bisa menyimpan file nya dalam folder projectnya jika ingin membuat seperti `localhost/index.js`

**Reply by taufik-nurrohman** [reply, marked as answer]
~~~ php
Hook::set('route', function ($content, $path) {
    if ('/index.js' === $path) {
        status(200);
        type('text/javascript');
        return content(LOT . D . 'y' . D . 'template' . D . 'index.js');
    }
    return $content;
}, 0);

Asset::set($url . '/index.js', 20);
~~~

~~~ php
Hook::set('content', function ($content) {
    return preg_replace('/\/index\.js\?v=\d+/', '/index.js');
});
~~~

**Reply by yuliakbard** [reply]
kalau untuk type data .txt `text/plain` ya kang?,

ini untuk `ads.txt`. tapi di console browser ada muncul error 403, walaupun sudah bisa diakses file nya

![image](https://github.com/mecha-cms/mecha/assets/96925342/f7e41d12-cb7a-44ba-9b1e-2f82bb388115)

**Reply by yuliakbard** [reply]
untuk javascript juga sama kang, ada error 403 nya

**Reply by taufik-nurrohman** [reply]
Tambahin `status(200)` sebelum `type('text/plain')`.

**Comment by taufik-nurrohman** [comment]
Konsepnya sama seperti yang pernah Saya jelaskan di artikel ini: https://mecha-cms.com/article/case-google-font

Cuma versi 3.x.x kan kelas `Route` sudah tidak ada digantikan sama kait. Kurang lebih seperti ini:

~~~ php
Hook::set('route', function ($content, $path) {
    if ('/manifest.json' === $path) {
        status(200);
        type('application/json');
        return To::JSON((array) State::get('x.f3h', true));
    }
    return $content;
}, 0);
~~~

**Reply by yuliakbard** [reply]
ohh, iya ada pembahasan cara buat rute seperti ini. thx kang

**Comment by yuliakbard** [comment]
kalau mau bikin route dengan `type('text/html')` gimana cara panggil file php nya?

``` php
Hook::set('route', function ($content, $path) {
    if ('/trending' === $path) {
        type('text/html');
        status(200);
        if (!empty($page)) {
        $out = file_get_contents(LOT . D . 'y' . D . 'model-1' . D . 'test.php');
        return strtr($content, ['</body>' => $out . '</body>']);
        }
    }
    return $content;
}, 0);
```

``` php
Hook::set('route', function ($content, $path) {
    if ('/trending' === $path) {
        type('text/html');
        status(200);
        function trending($content) {
            extract($GLOBALS, EXTR_SKIP);
            if (!empty($page)) {
                $out  = '<h2>Trending</h2>';
                return strtr($content, ['</body>' => '</body>' . $out]);
            }
            return $content;
        }
        Hook::set('content', __NAMESPACE__ . "trending", 0.5);
    }
    return $content;
}, 0);
```

``` php
Hook::set('route', function ($content, $path) {
    if ('/trending' === $path) {
        type('text/html');
        status(200);
        function trending($content) {
            extract($GLOBALS, EXTR_SKIP);
            if (!empty($page)) {
                return layout::test(); // Ambil test.php
            }
            return $content;
        }
        Hook::set('content', __NAMESPACE__ . "trending", 0.5);
    }
    return $content;
}, 0);

```

``` php
Hook::set('route', function ($content, $path) {
    if ('/trending' === $path) {
        type('text/html');
        status(200);
        return Hook::set('content', function () {
            $content = layout::test();
            return $content;
        });
    }
    return $content;
}, 0);
```


cuma kode diatas yang terpikirkan sama ane kang, bantu sempurnakan kang. untuk penambahan `State::set()` dibagian mana nya kang?

**Reply by taufik-nurrohman** [reply]
Daripada nge-hook satu-satu mending cobain ekstensi [Route](https://github.com/mecha-cms/x.route) mas.

**Reply by yuliakbard** [reply]
cara tampilin $content, masih gak mudeng ane kang, butuh contoh kang. minimal cara nampilin element

![image](https://github.com/mecha-cms/mecha/assets/96925342/0f190ac4-90f0-4efc-9f50-a6200f3b22a7)
 
pas di return pake `return layout::test();` state di <html> nya hilang. 😅

**Reply by taufik-nurrohman** [reply]
Coba `return ['test', [], 200];` mas.

**Reply by yuliakbard** [reply]
outputnya sama, dengan pake kode ini.

```
Hook::set('route', function ($content, $path) {
    if ('/trending' === $path) {
        type('text/html');
        status(200);
            $content = "Test Halaman";
            return $content;
    }
    return $content;
}, 0);
```

di Halaman Route di coba pada halaman `http://localhost/hero/route/bio.php`

```
return function () {
    status(200);
    $abc = "XXXX XXXX XXXX";
    return $abc;
};
```
ataupun seperti ini.
```
return function () {
    return ['test', [], 200];
};
```
![image](https://github.com/mecha-cms/mecha/assets/96925342/45652f43-51d1-4060-9fea-bffbb837ada4)

**Reply by taufik-nurrohman** [reply]
Mas ada berkas `.\lot\y\*\test.php` kan? Soalnya `['test', [], 200]` itu nanti akan menjadi `Layout::get('test', [], 200)`.

**Reply by yuliakbard** [reply]
ada kang. bahkan ane coba seperti ini juga. dan hasilnya juga sama

```
return function () {
$content = "Ini Content";
    return [$content, [], 200];
};
```

**Reply by taufik-nurrohman** [reply]
Nah output-nya ditulis di dalem berkas `.\lot\y\*\test.php` mas.

**Reply by yuliakbard** [reply]
isinya `test.php` harus seperti ini rupanya ya kang:
```
<?=self::enter();?>
<h1><?= $page->title; ?></h1>
<h2>Saya di Halaman Bio</h2>
<?=self::exit();?>
```

Hasilnya:
![image](https://github.com/mecha-cms/mecha/assets/96925342/542908a4-2da4-489c-a995-0f2889036cdd)

Tapi muncul pertanyaan baru, kan ane ikutin petunjuk yang ada di extension Route.

di dalam file `route/bio.php` ini isinya seperti ini.

```
<?php
return function () {
    State::set([
        'has' => [
            'next' => false,
            'page' => true,
            'pages' => false,
            'part' => false,
            'prev' => false
        ],
        'is' => [
            'error' => false,
            'page' => true,
            'pages' => false
        ]
    ]);
    $GLOBALS['page'] = $page = new Page(LOT . D . 'page' . D . 'bio' . '.' . 'page');
    $GLOBALS['t'][] = $page->title;
    return ['test', [], 200];
};
```

Untuk $page->title nya masih belum muncul. 

Bagian ininya kayaknya ane masih salah.
`$GLOBALS['page'] = $page = new Page(LOT . D . 'page' . D . 'bio' . '.' . 'page');`

**Reply by yuliakbard** [reply]
seperti ini bisa kang tampil `$page->title nya`

```
    $GLOBALS['page'] = $page = new Page(null, [
            'title' => 'Bio'
        ]);
```

**Comment by bungfrangki** [comment]
> ```
> Hook::set('route', function ($content, $path) {
>     if ('/index.js' === $path) {
>         status(200);
>         type('text/javascript');
>         return content(LOT . D . 'y' . D . 'template' . D . 'index.js');
>     }
>     return $content;
> }, 0);
> 
> Asset::set($url . '/index.js', 20);
> ```
> 
> ```
> Hook::set('content', function ($content) {
>     return preg_replace('/\/index\.js\?v=\d+/', '/index.js');
> });
> ```

kalau mau nulis kode di atas dimana, mas?

**Reply by taufik-nurrohman** [reply]
Di dalam berkas `index.php` layout yang aktif. Kalau di WordPress biasanya akan masuk ke berkas `functions.php`. Konsepnya mirip.

**Reply by bungfrangki** [reply]
iya mas, sudah ketemu. keren juga yaa, kita bisa buat path sesuka hati

**Reply by yuliakbard** [reply]
suhu suhu blogger, mulai nimbrung dimari. hihihi. ayo sebarin mecha
